### PR TITLE
[feature] [null support # 10] Add null support in all transform function and pass the bitmap to upstream

### PR DESCRIPTION
**Pull request by 61yao**
Pass the null bitmap up from transform function to transform blockValueSet.

We only do the null related calculation when enableNullHandling is set to true due to performance reason.

Null support issue:
https://github.com/apache/pinot/issues/10252

Design doc:
https://docs.google.com/document/d/1w2eBuZrS73xE2JIw8MYlySy5pNaX9iE8kZr_YAL20CM/edit

### CODE REVIEW

**Review by shenyu0127 — COMMENTED** [review]
Fix linter and ensure all tests pass?

**Comment by KKcorps** [comment]
Seems like another rebase needs to be done to resolve conflicts from previous PR

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CoalesceTransformFunction.java [hunk 1]
```diff
@@ -399,4 +390,22 @@ public String[] transformToStringValuesSV(ValueBlock valueBlock) {
...
+  @Override
+  public RoaringBitmap getNullBitmap(ValueBlock valueBlock) {
+    RoaringBitmap[] nullBitMaps = getNullBitMaps(valueBlock, _transformFunctions);
+    RoaringBitmap bitmap = nullBitMaps[0];
+    for (int i = 1; i < nullBitMaps.length; i++) {
+      RoaringBitmap curBitmap = nullBitMaps[i];
+      if (bitmap != null && curBitmap != null) {
+        bitmap.and(curBitmap);
+      } else {
+        bitmap = null;
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CoalesceTransformFunction.java:403]: we can break the loop when setting bitmap to null

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/DateTimeConversionTransformFunction.java [hunk 2]
```diff
@@ -106,8 +107,8 @@ public void init(List<TransformFunction> arguments, Map<String, ColumnContext> c
 
     _dateTimeTransformer = DateTimeTransformerFactory.getDateTimeTransformer(
         ((LiteralTransformFunction) arguments.get(1)).getStringLiteral(),
-            ((LiteralTransformFunction) arguments.get(2)).getStringLiteral(),
-            ((LiteralTransformFunction) arguments.get(3)).getStringLiteral());
+        ((LiteralTransformFunction) arguments.get(2)).getStringLiteral(),
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/DateTimeConversionTransformFunction.java:110]: nit: unnecessary formatting

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/GreatestTransformFunction.java [hunk 3]
```diff
@@ -74,6 +135,35 @@ public float[] transformToFloatValuesSV(ValueBlock valueBlock) {
...
   }
 
+  @Override
+  public Pair<float[], RoaringBitmap> transformToFloatValuesSVWithNull(ValueBlock valueBlock) {
+    int numDocs = valueBlock.getNumDocs();
+    initFloatValuesSV(numDocs);
+    Pair<float[], RoaringBitmap> values = _arguments.get(0).transformToFloatValuesSVWithNull(valueBlock);
+    System.arraycopy(values.getLeft(), 0, _floatValuesSV, 0, numDocs);
+    RoaringBitmap nullBitmap = values.getRight();
+    for (int i = 1; i < _arguments.size(); i++) {
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/GreatestTransformFunction.java:145]: Can we extract out this logic in a method. Seems like the only difference is the datatype of the array.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/InTransformFunction.java [hunk 4]
```diff
@@ -56,6 +56,8 @@ public String getName() {
 
   @Override
   public void init(List<TransformFunction> arguments, Map<String, ColumnContext> columnContextMap) {
+    // TODO: Proper support for null in groovy functions.
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/InTransformFunction.java:59]: Can you elaborate this TODO a bit more. What's missing to provide proper support?

File: pinot-core/src/main/java/org/apache/pinot/core/operator/docvalsets/TransformBlockValSet.java [hunk 5]
```diff
@@ -46,47 +45,23 @@ public class TransformBlockValSet implements BlockValSet {
...
+      ExpressionContext expression, boolean isNullHandlingEnabled) {
     _valueBlock = valueBlock;
     _transformFunction = transformFunction;
     _expression = expression;
+    _isNullHandlingEnabled = isNullHandlingEnabled;
   }
 
   @Nullable
   @Override
   public RoaringBitmap getNullBitmap() {
```
Comment (shenyu0127) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/docvalsets/TransformBlockValSet.java:64]: What if `getNullBitmap` is called before any of the `getXXValuesSV` is called?

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/NullHandlingIntegrationTest.java [hunk 6]
```diff
@@ -299,4 +296,189 @@ public void testNullLiteralSelectionOnlyBroker()
     assertEquals(rows.size(), 1);
     assertEquals(rows.get(0).get(0).asText(), "null");
   }
+
+  @Test
+  public void testNullSelectionOnlyTransform()
```
Comment (shenyu0127) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/NullHandlingIntegrationTest.java:301]: nit: break down the test into multiple tests so that we test behaviors separately.

File: pinot-core/src/main/java/org/apache/pinot/core/common/RowBasedBlockValueFetcher.java
Comment (shenyu0127) [inline pinot-core/src/main/java/org/apache/pinot/core/common/RowBasedBlockValueFetcher.java:1]: Please file an issue to create a unit test file for this file as well as TransformBlock and TransformBlockValSet.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/docvalsets/TransformBlockValSet.java (hunk 5 above)
Comment (shenyu0127) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/docvalsets/TransformBlockValSet.java:64]: We only need to address this comment, the other comments are nit-picks and optional.

**Review by shenyu0127 — COMMENTED** [review]
Hi Yao, are you able to complete this PR? 

Only one comment needs to be addressed, the rest looks good to me.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/docvalsets/TransformBlockValSet.java [hunk 7]
```diff
@@ -118,111 +93,181 @@ public int[] getDictionaryIdsSV() {
   public int[] getIntValuesSV() {
     try (InvocationScope scope = Tracing.getTracer().createScope(TransformBlockValSet.class)) {
       recordTransformValues(scope, DataType.INT, true);
-      return _transformFunction.transformToIntValuesSV(_valueBlock);
+      if (!_isNullHandlingEnabled) {
+        return _transformFunction.transformToIntValuesSV(_valueBlock);
+      }
+      Pair<int[], RoaringBitmap> result = _transformFunction.transformToIntValuesSVWithNull(_valueBlock);
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/docvalsets/TransformBlockValSet.java:99]: Should we also add `getIntValuesSVWithNull()` etc so that it is consistent with transform function behavior (separate the APIs with or without null handling)

**Comment by shenyu0127** [comment]
@61yao will not continue working on this PR; I will patch this PR and finish it.
